### ISSUE TITLE
Eyeglass modules not discovered on Windows

### ISSUE THREAD

**Issue by JohnRiv** [sass-eyeglass/eyeglass#128]
Testing on Windows 10, I noticed that my eyeglass modules that import automatically on my Mac are not imported on Windows. I've narrowed it down to [this path comparison](https://github.com/sass-eyeglass/eyeglass/blob/e8f4161924e24e8284ff2f0733ade9d9cbb15deb/lib/util/eyeglass_modules.js#L251):

```js
// if the module path matches the search path, push it onto our results
if (mod.path === dir) {
  branches.push(mod);
}
```

On a Mac, these are identical, e.g.:
mod.path: `/Users/johnriv/projects/eyeglass-test`
dir: `/Users/johnriv/projects/eyeglass-test`

However on Windows, they're different:
mod.path: `C:\Users\IEUser\Documents\eyeglass-test`
dir: `C:/Users/IEUser/Documents/eyeglass-test`

**Comment by eoneill** [comment]
Thanks for reporting this. I suspected this last night as I was working on some cleanup. We're not very consistent with the way we handle uris and file paths and I started writing some helpers to normalize this.

I'm cleaning stuff up for assets right now, and should be able to work this in.

**Comment by JohnRiv** [comment]
Sounds great, thanks. Let me know when the code is pushed up and I can test it out.

**Comment by eoneill** [comment]
I've got most things working now on Windows (tested on Win10) with the latest changes in my branch: https://github.com/eoneill/eyeglass/tree/assets-refactor

I still have a bit more cleanup to do, but things should work on Windows soon.

Will update when I submit a pull request.

**Comment by JohnRiv** [comment]
Thanks for the update @eoneill. I tried out your progress so far and it works on my basic Eyeglass test project. With eyeglass-spriting though, I'm running into this error on Windows 10:

```
C:\Users\IEUser\Documents\eyeglass-test>grunt css
Running "sass:dist" (sass) task
>> error in C function eyeglass-asset-uri: Asset not found: spritemaps\icon-sprite-map.png
>>
>> Backtrace:
>>      node_modules/eyeglass/sass/assets.scss:48, in function `eyeglass-asset-uri`
>>      node_modules/eyeglass/sass/assets.scss:48, in function `asset-uri`
>>      node_modules/eyeglass/sass/assets.scss:53, in function `asset-url`
>>      node_modules/eyeglass-spriting/sass/index.scss:17, in function `sprite-url`
>>      node_modules/eyeglass-spriting/sass/index.scss:25, in mixin `sprite-background`
>>      assets/sass/_sprites.scss:17
>>   Line 48  Column 11  node_modules\eyeglass\sass\assets.scss
Warning:  Use --force to continue.
```

Comparing a successful run on my Mac vs the error on Windows 10, I noticed [this code from the `resolveAssetDefaults` function in lib/assets/Assets.js](https://github.com/eoneill/eyeglass/blob/dea7e6ca322bf4c8cd57fa4195ba544df8ee256e/lib/assets/Assets.js#L145-L158) is what is succeeding in OSX and failing in Windows 10:

```js
function resolveAssetDefaults(registeredAssetsMap, relativePath) {
  registeredAssetsMap = this.sassUtils.handleEmptyMap(registeredAssetsMap);
  this.sassUtils.assertType(registeredAssetsMap, "map");

  var registeredAssets = this.sassUtils.castToJs(registeredAssetsMap);

  var appAssets = registeredAssets.coerce.get(null);
  if (appAssets) {
    // XXX sassUtils.assertType(appAssets, "map");
    var appAsset = appAssets.coerce.get(relativePath);
    if (appAsset) {
      return appAsset;
    }
  }
  ...
```

In this case, the `relativePath` argument passed in is `spritemaps/icon-sprite-map.png` (for both Mac & Windows, which I guess makes sense since that comes from eyeglass-spriting and that isn't doing the normalization). The `registeredAssets` on Mac contains these relevant sassKeys:
```js
     'string:images/spritemaps/icon-sprite-map.png' => SassString {},
     'string:spritemaps/icon-sprite-map.png' => SassString {},
```
and these relevant sassValue:
```js
     'string:images/spritemaps/icon-sprite-map.png' => SassMap {},
     'string:spritemaps/icon-sprite-map.png' => SassMap {},
```
so it ends up assigning and returning an `appAsset`. 

However, in Windows 10, the related sassKeys are:
```js
     'string:images/spritemaps/icon-sprite-map.png' => SassString {},
     'string:spritemaps\\icon-sprite-map.png' => SassString {} },
```
and the related sassValues are:
```js
     'string:images/spritemaps/icon-sprite-map.png' => SassMap {},
     'string:spritemaps\\icon-sprite-map.png' => SassMap {} },
```
so no `appAsset` is assigned nor returned.

Note that the `\\` may actually just be a `\` and it's appearing that way due to how I'm logging out the values (I noticed that when originally debugging this issue that a `\` would be logged out as `\\`), but either way, you can see the path delimiter mismatch there.

If there is any further information I can provide to help you out just let me know. Thanks.

**Comment by eoneill** [comment]
Opened another issue against `eyeglass-spriting` for tracking.

**Comment by eoneill** [comment]
The `eyeglass-spriting` issues *might* be addressed with my latest commit in #129, but I wouldn't count on it.

I think there are still some changes that need to happen on they eyeglass-spriting side, which I'll look into after this PR lands.

I'll test this PR against eyeglass-spriting on Windows later tonight.

**Comment by JohnRiv** [comment]
Thanks for your work on this. Here's what I'm seeing with the [latest in eyeglass-spriting](https://github.com/sass-eyeglass/eyeglass-spriting/tree/ac9a09218579678eb3128783303d38dbe03b49ee) and your latest (currently 3rd) commit to eyeglass PR #129:

- Project with just eyeglass, using grunt-sass:
	- **OSX:** works!
	- **Win10:**  works!

- Project with eyeglass & eyeglass-spriting, using grunt-sass (the "Base project" in the other examples below):
	- **OSX:** works!
	- **Win10:** `Fatal error: First argument should be a number.`
	- **Reference Project Commit:** https://github.com/JohnRiv/eyeglass-spriting-grunt/commit/5b13cf53584f75ffe0bdb114fc2ccba0df466763

- Add an Eyeglass module (such as [Bourbon](https://github.com/thoughtbot/bourbon)) to the Base project and `@import` it in a SCSS file (don't even use a function or mixin from the library. Note I verified it installed to the node_modules directory)
	- **OSX:** `Fatal error: Path must be a string. Received undefined`
	- **Win10:**
```bash
>> Error: Could not import bourbon from any of the following locations:
>>   C:\Users\IEUser\Documents\eyeglass-spriting-grunt\assets\sass\bourbon.scss
>>   C:\Users\IEUser\Documents\eyeglass-spriting-grunt\assets\sass\bourbon.sass
>>   C:\Users\IEUser\Documents\eyeglass-spriting-grunt\assets\sass\bourbon.css
>>   C:\Users\IEUser\Documents\eyeglass-spriting-grunt\assets\sass\_bourbon.scss
>>   C:\Users\IEUser\Documents\eyeglass-spriting-grunt\assets\sass\_bourbon.sass
>>   C:\Users\IEUser\Documents\eyeglass-spriting-grunt\assets\sass\_bourbon.css
>>   C:\Users\IEUser\Documents\eyeglass-spriting-grunt\assets\sass\bourbon\index.scss
>>   C:\Users\IEUser\Documents\eyeglass-spriting-grunt\assets\sass\bourbon\index.sass
>>   C:\Users\IEUser\Documents\eyeglass-spriting-grunt\assets\sass\bourbon\index.css
>>   C:\Users\IEUser\Documents\eyeglass-spriting-grunt\assets\sass\bourbon\_index.scss
>>   C:\Users\IEUser\Documents\eyeglass-spriting-grunt\assets\sass\bourbon\_index.sass
>>   C:\Users\IEUser\Documents\eyeglass-spriting-grunt\assets\sass\bourbon\_index.css
>>   Line 5  Column 9  assets\sass\styles.scss
```
**Reference Project Commit:** https://github.com/JohnRiv/eyeglass-spriting-grunt/commit/ba5cfb5ed3d7d33232400cafe1c054ca86f2f45c

- Add a Sass library (such as [include-media](http://include-media.com/)) to the Base project via includePaths and then `@import` it in a SCSS file and call a mixin or function
	- **OSX:** works!
	- **Win10:** Fatal error: First argument should be a number.

So it appears that:
1. A basic eyeglass project works in OSX and Windows
2. There are still some issues with eyeglass-spriting in Windows
3. Other eyeglass modules (such as Bourbon) have issues in both OSX & Windows

When I have a chance I'll try to debug those, but I wanted to make sure you had this info.

**Comment by eoneill** [comment]
Thanks for the thorough report and providing a github repo to reproduce. This was incredibly helpful and I was able to debug the issues. If you use the latest at `eoneill/eyeglass#assets-refactor-clean`, everything should work on Windows, OSX, and Linux now.

**Comment by JohnRiv** [comment]
You're awesome @eoneill, thanks for your late-night dedication to this. Hope you've been able to catch up on some sleep since then. I ran through all the cases and they're all looking good to me except the last one:

- Add a Sass library (such as [include-media](http://include-media.com/)) to the Base project via includePaths and then `@import` it in a SCSS file and call a mixin or function
  - OSX: `Fatal error: Path must be a string. Received undefined`
  - Win10: `no mixin named media`

Here's the reference project commits for that:
- Adding include-media to the project via includePaths: https://github.com/JohnRiv/eyeglass-spriting-grunt/commit/6f2203dedcb45bb2f609826fa8dbd5a98d325b6c
- Actually using the media mixin from include-media: https://github.com/JohnRiv/eyeglass-spriting-grunt/commit/39c9695272665a0ef588233779487c2adedf0243
That first commit works, it's the second that throws the error. You'll need both to see it (just grab the latest from [JohnRiv/eyeglass-spriting-grunt#assets-refactor](https://github.com/JohnRiv/eyeglass-spriting-grunt/commits/assets-refactor))

Interesting that it worked in OSX originally but now doesn't, although the error message in Windows is more straightforward now at least.

**Comment by eoneill** [comment]
If you add `@import "include-media"`, the `@include media(...)` call should work.

The error on OSX is definitely pretty useless. After debugging it a bit, it looks like the `Path must be a string...` error is originating in `grunt-sass` on this line:
https://github.com/sindresorhus/grunt-sass/blob/v1.1.0/tasks/sass.js#L28
In this case, `err.file` is `undefined`, and `path.resolve` throws the above error.

This might be a deeper issue with node-sass and/or libsass as the error object coming back does not contain `file`.

**Comment by JohnRiv** [comment]
Derp, that would certainly explain it ;-) Thanks. I'll look more into grunt-sass and node-sass to see how a more useful error can be thrown. 

Thanks again, I appreciate your hard work on this.

**Comment by JohnRiv** [comment]
Looks like the next release of grunt-sass should address that useless error message: https://github.com/sindresorhus/grunt-sass/pull/252

**Comment by Snugug** [comment]
I'm currently running in to this issue; eyeglass modules are not being discovered on Windows computers and `node-sass` is erroring out because it cannot find the required imports. Any ETA on this? It looks like it's been open for quite a few months. Using the latest version of Eyeglass

**Comment by iammikecohen** [comment]
Same. Does anyone have a workaround for this? I've tried changing node-sass and eyeglass versions and was not able to find a combination that works.

**Comment by eoneill** [comment]
This should be officially resolved with eyeglass @ 0.9.0. Per #137.
